### DISCUSSION TITLE
Appropriate way setting borg.drones-directory for symlinked .emacs.d

### DISCUSSION THREAD

**Discussion by gwbrck** [emacscollective/borg#117] [Support] [answered]
Hi 😊

in my configurations I have a symlink from `~/dotfiles/something/.emacs.d` to `~/.emacs.d.`

What value would I have to set borg.drones-directory to if I want to start emacs normally via `emacs` without `-Q -l`? 
Git submodules should be fine in the folder "~/dotfiles/something/.emacs.d/lib" - right? 

Is ist even possible to use borg this way? 

Many thanks!

EDIT
Maybe I should have mentioned that ~/dotfiles/ is the git repository and not ~/dotfiles/something/.emacs.d/.
So it's ~/dotfiles/.git not ~/dotfiles/something/.emacs.d/.git.

**Comment by tarsius** [comment, marked as answer]
Just give it a try my best guess is that it just works.

**Reply by gwbrck** [reply]
Maybe I should have mentioned that `~/dotfiles/` is the git repository and not `~/dotfiles/something/.emacs.d/`. 
So it's `~/dotfiles/.git` not `~/dotfiles/something/.emacs.d/.git`.
Without configuration of borg.drones-directory borg would add submodules under `~/dotfiles/lib`. 

I tried to set borg.drones-directory to `something/.emacs.d/lib` but when calling `make bootstrap` there is an build error: 

"No such file or directory" "/Users/gwbrck/dotfiles/something/.emacs.d/something/.emacs.d") 

The path seems to be doubled somehow... 

Thanks for your quick response!

**Reply by tarsius** [reply]
That's a relevant little detail indeed. :grinning: 

Someone else wanted to do the same thing before and we talked about it in this issue tracker. Please go looking for that conversation; it would take me just as long to dig it up for you. (And then tell me which it is so I can add the `faq` label.)

**Reply by gwbrck** [reply]
https://github.com/emacscollective/borg/issues/36 - but I'm not sure that's what you meant.
Maybe this too https://github.com/emacscollective/borg/issues/9

**Reply by tarsius** [reply]
I though there was something more recent, which included a working configuration, but I might misremember.

**Reply by tarsius** [reply]
Could you please edit your posts so far to remove the `(...)` obfuscation? That makes things ambiguous, for example when you say:

> Maybe I should have mentioned that `~/dotfiles/(...)` is the git repository 

Are we then talking about `~/dotfiles/.git` or `~/dotfiles/something/.git`? Not sure if that even matters here, but its always nice to operate with as few uncertainties as possible.

**Reply by tarsius** [reply]
> I tried to set borg.drones-directory to `(...)/.emacs.d/lib`

Try with an absolute path instead. Try both paths, the one beginning with `~/dotfiles` and the one beginning with `~/.emacs.d`.

**Reply by gwbrck** [reply]
So the doubled path seems to be fixed, when using an absolute path. Also just not setting borg.drones-directory and putting the Makefile under `~/dotfiles/something/.emacs.d/Makefile` works.
But I have still errors when run make bootstrap and I just can't figure out the error message.

```
=== Running 'make build' ===

Initializing drones...done (2 drones in 0.007s, 33 skipped)
Debugger entered--Lisp error: (args-out-of-range "160000 5b7c140f5b768141ce749c3254fe00abd593128a 0\11..." 50 75)
  substring("160000 5b7c140f5b768141ce749c3254fe00abd593128a 0\11..." 50 75)
  (string-equal (substring line 50 offset) prefix)
  (and (string-equal (substring line 50 offset) prefix) (list (substring line offset)))
  (closure ((offset . 75) (prefix . "Packs/emacs/.emacs.d/lib/") (include-variables) compilation-mode-font-lock-keywords git-commit-mode-map t) (line) (and (string-equal (substring line 50 offset) prefix) (list (substring line offset))))("160000 5b7c140f5b768141ce749c3254fe00abd593128a 0\11...")
  cl-mapcan((closure ((offset . 75) (prefix . "Packs/emacs/.emacs.d/lib/") (include-variables) compilation-mode-font-lock-keywords git-commit-mode-map t) (line) (and (string-equal (substring line 50 offset) prefix) (list (substring line offset)))) ("160000 2f57074fc9f67128b30ce32949130a8dbda0b098 0\11..." "160000 5056e635ecf406e65d7d28651bab55600dd18741 0\11..." "160000 809e7d77fec834597f495fec737c59292a10025b 0\11..." "160000 9a2fab9b212a76b01729f7f752c09e89b47a9d11 0\11..." "160000 39f06b873a8449af8ff6a3eee716d3da14d63a76 0\11..." "160000 c36871b2a44b59761387f4972c617b44dcec5e75 0\11..." "160000 87b0f86da3dffef63b42845049c648b5d90f1c4d 0\11..." "160000 45c9b2faff4785539a0d0c655440c2465fed985a 0\11..." "160000 dac55912be3642a954ef0891b63a3f4cb1db5ce4 0\11..." "160000 e8e5303f9ee3d8dc327c97891eaa1257ba5d4eee 0\11..." "160000 36c610a9717cc9ec426a07c8e6bf3b3abcb139d6 0\11..." "160000 9de4ceb9314aa59a726d3f0b31678e9d955a353e 0\11..." "160000 0e25e0e98990803e95c7851236e43b3ee934d443 0\11..." "160000 ee101462d127ed6a5561ce9ce92bfded87d7d478 0\11..." "160000 787a7a8d4444e58467d6b3d4b88a497e7d494643 0\11..." "160000 b42e0cfe22ad3a0cac35cf14315e9dc22cea2ec9 0\11..." "160000 f91d80973f80025d4ed00380f2e06c669dfda49d 0\11..." "160000 01a395df00adbb57813e852f2e83ffb12375f1b6 0\11..." "160000 f3ef5920725d719ed23195a457a86015826d3e1d 0\11..." "160000 a6b823b8d0a8959e81876fd4857ba822495c028a 0\11..." "160000 256702dd1432894b3607d3de6cd660863b331818 0\11..." "160000 aeb933272e72617f7c4d35e1f003be16836b948d 0\11..." "160000 0321f7ca74cd56297b3b29f667a752854e1c2eee 0\11..." "160000 da167c51e9fd167a48d06c7c0ee8e3ac7abd9718 0\11..." "160000 fb6f71a0288b5aec5723b6118aa0ca02e2cc9a23 0\11..." "160000 88f09de2a06f95bd1933b950ec2b66671ae36fbd 0\11..." "160000 5509c25ac1b09617848adbb517c8443580a82874 0\11..." "160000 5b7c140f5b768141ce749c3254fe00abd593128a 0\11..." "160000 be50b05052da5d6851b824c1c5a14f9f7aaf198f 0\11..." "160000 a411ef3e0992d4839f0732ebeb9823024afaaaa8 0\11..." "160000 1a9264bc661b3d52756916bf9ec3f41687d64db2 0\11..."))
  (let* ((default-directory borg-top-level-directory) (prefix (file-relative-name borg-drones-directory)) (offset (+ (length prefix) 50))) (cl-mapcan #'(lambda (line) (and (string-equal (substring line 50 offset) prefix) (list (substring line offset)))) (process-lines "git" "submodule--helper" "list")))
  (if include-variables (let (alist) (let ((--dolist-tail-- (and (file-exists-p borg-gitmodules-file) (process-lines "git" "config" "--list" "--file" borg-gitmodules-file)))) (while --dolist-tail-- (let ((line (car --dolist-tail--))) (if (string-match "\\`submodule\\.\\([^.]+\\)\\.\\([^=]+\\)=\\(.+\\)\\'" line) (progn (let* ... ... ... ...))) (setq --dolist-tail-- (cdr --dolist-tail--))))) (cl-sort alist #'string< :key #'car)) (let* ((default-directory borg-top-level-directory) (prefix (file-relative-name borg-drones-directory)) (offset (+ (length prefix) 50))) (cl-mapcan #'(lambda (line) (and (string-equal (substring line 50 offset) prefix) (list (substring line offset)))) (process-lines "git" "submodule--helper" "list"))))
  borg-drones()
  (let ((drones (borg-drones))) (if (member "org" drones) (progn (setq drones (cons "org" (delete "org" drones))))) (let ((--dolist-tail-- drones)) (while --dolist-tail-- (let ((drone (car --dolist-tail--))) (if (or (equal (borg-get drone "disabled") "true") (not (file-exists-p ...)) (and quick (borg-get-all drone "build-step"))) nil (let ((--dolist-tail-- ...)) (while --dolist-tail-- (let ... ... ...)))) (setq --dolist-tail-- (cdr --dolist-tail--))))) (let ((--dolist-tail-- drones)) (while --dolist-tail-- (let ((drone (car --dolist-tail--))) (message "\n--- [%s] ---\n" drone) (cond ((equal (borg-get drone "disabled") "true") (message "Skipped (Disabled)")) ((not (file-exists-p ...)) (message "Skipped (Missing)")) ((and quick (borg-get-all drone "build-step")) (message "Skipped (Expensive to build)")) (t (borg-build drone))) (setq --dolist-tail-- (cdr --dolist-tail--))))))
  borg-batch-rebuild()
  command-line-1(("--eval" "(require 'comp nil t)" "--load" "subr-x" "--eval" "(setq byte-compile-warnings '(not docstrings))" "--eval" "(fset 'original-message (symbol-function 'message)..." "--eval" "(fset 'message (lambda (format &rest args) (unless..." "-L" "lib/borg/" "--load" "borg" "--funcall" "borg-initialize" "--funcall" "borg-batch-rebuild"))
  command-line()
  normal-top-level()

make[1]: *** [build] Error 255
make: *** [bootstrap] Error 2
```

**Reply by gwbrck** [reply]
So I think the problems come up because I don't know what I'm doing :D 

So my original problem. The one from this question is actually answered with your first answer. I have recreated the file structure as I want it with `~/dotfiles/something/.emacs.d/`, in a new repository. 
And it just works.

I think my real problem was that I was using my old dotfiles repository which was already filled with submodules. 
I will take a look at this https://github.com/emacscollective/borg/issues/92. 

Sorry for bothering you with this. 🙈

**Reply by gwbrck** [reply]
In fact the problem is other submodules that have nothing to do with emacs. These submodules are not in `~/dotfiles/something/.emacs.d/lib` but somewhere else. Is there any way that borg can ignore these submodules?
I tried setting `submodule.DRONE.disabled` for them to `true`. But it didn't work...

**Reply by tarsius** [reply]
> In fact the problem is other submodules that have nothing to do with emacs.

That should be fixed now.

**Reply by tarsius** [reply]
Uh, actually the fix is broken, but I should have a working one by tomorrow.

**Reply by tarsius** [reply]
Pushed the fixed version.

**Reply by gwbrck** [reply]
It works. Awesome!
Thank you 🙌🏻

**Reply by gwbrck** [reply]
I tried it a bit and everything seems to work👌🏻 however, when i run `make all` an error comes up at the end where the path is displayed incorrectly... 
```
--- [init.el] ---

Loading /Users/gwbrck/dotfiles/something/.emacs.d/init.el (source)...
Initializing drones...done (66 drones in 0.072s)
Warning (evil-collection): `evil-want-keybinding' was set to nil but not before loading evil.

Make sure to set `evil-want-keybinding' to nil before loading evil or evil-collection.

See https://github.com/emacs-evil/evil-collection/issues/60 for more details.
Clearing removed files...
Clearing removed files...done
Processing modified files...
Processing modified files...done
[yas] Prepared just-in-time loading of snippets (but no snippets found).
Compiling /Users/gwbrck/init.el...
Debugger entered--Lisp error: (file-missing "Opening input file" "No such file or directory" "/Users/gwbrck/init.el")
  insert-file-contents("/Users/gwbrck/init.el")
  byte-compile-file("/Users/gwbrck/init.el")
  byte-recompile-file("/Users/gwbrck/init.el" t 0)
  borg-batch-rebuild-init()
  borg-batch-rebuild()
  command-line-1(("--eval" "(require 'comp nil t)" "--load" "subr-x" "--eval" "(setq byte-compile-warnings '(not docstrings))" "--eval" "(fset 'original-message (symbol-function 'message)..." "--eval" "(fset 'message (lambda (format &rest args) (unless..." "-L" "lib/borg/" "--load" "borg" "--funcall" "borg-initialize" "--funcall" "borg-batch-rebuild"))
  command-line()
  normal-top-level()
```
While /Users/gwbrck/dotfiles/something/.emacs.d/init.el is right Users/gwbrck/init.el is wrong.

**Reply by tarsius** [reply]
The `user-dir` branch should fix that, please give it a try. I want to investigate why the borg variant wasn't used in those cases in the first place before merging into `master`.

**Reply by gwbrck** [reply]
I used `git checkout user-dir`. Still the error occurs when I run `make all`. But it's just with `make build init`. So maybe `INIT_FILES` wasn't set properly?

**Reply by gwbrck** [reply]
when I run `make lib/org`, for example, everything works.

**Reply by tarsius** [reply]
>  But it's just with `make build init`.

There is no `init` target. There is a `tangle-init` target, which just depends on the `init.el` target, which generates `init.el` from `init.org`. `all` and `build` do not depend on `init.el` or `tangle-init`.

> Still the error occurs when I run `make all`.

`all` depends on `build`, which does `emacs ... --funcall borg-batch-rebuild $(INIT_FILES)`. `borg-batch-rebuild` calls `borg-batch-rebuild-init`, which uses `command-line-args-left` (known to `make` as `$(INIT_FILES)`).

> So maybe `INIT_FILES` wasn't set properly?

Yes indeed. What did you set it too? :grin:
You should set it in `.emacs.d/Makefile` and the paths should be absolute (or relative to `.emacs.d` I think). Though not setting it at all should work too I think. Add debug statements to `borg-batch-rebuild-init` to figure out what is going on.
